File: modules/day-04-reliable/demos/unit-test.R
# Day 4 demo: minimal testthat checks for the smoothing helper.

library(testthat)

source("silent_break.R")

test_that("smooth_time_series returns expected values for typical data", {
  values <- data.frame(
    Sample_ID = c("Sample_1", "Sample_2"),
    Condition = c("A", "B"),
    Sex = c("Male", "Female"),
    `0` = c(0, 10),
    `1` = c(0, 10),
    `2` = c(6, 16),
    `3` = c(6, 16),
    check.names = FALSE
  )

  expected <- tibble::tibble(
    Condition = c("A", "B", "A", "B", "A", "B", "A", "B"),
    Sex = c("Male", "Female", "Male", "Female", "Male", "Female", "Male", "Female"),
    Time = c(0, 0, 1, 1, 2, 2, 3, 3),
    Value = c(0, 10, 2, 12, 4, 14, 6, 16)
  ) |>
    dplyr::arrange(Condition, Sex, Time)

  actual <- smooth_time_series(values, window = 3) |>
    dplyr::arrange(Condition, Sex, Time)

  expect_equal(actual, expected)
})

test_that("smooth_time_series fails when timepoints are not evenly spaced", {
  values <- data.frame(
    Sample_ID = c("Sample_1", "Sample_2"),
    Condition = c("A", "B"),
    Sex = c("Male", "Female"),
    `0` = c(0, 10),
    `2` = c(6, 16),
    `3` = c(6, 16),
    check.names = FALSE
  )

  expect_error(
    smooth_time_series(values, window = 3),
    "Timepoints are not evenly spaced."
  )
})

File: modules/day-04-reliable/demos/unit-test.py
"""Day 4 demo: minimal pytest checks for the smoothing helper."""

import pandas as pd
import pytest

from slient_break import _smooth_time_series


def test_smooth_time_series_returns_expected_values_for_typical_data():
    values = pd.DataFrame(
        {
            "Sample_ID": ["Sample_1", "Sample_2"],
            "Condition": ["A", "B"],
            "Sex": ["Male", "Female"],
            "0": [0.0, 10.0],
            "1": [0.0, 10.0],
            "2": [6.0, 16.0],
            "3": [6.0, 16.0],
        }
    )

    expected = pd.DataFrame(
        {
            "Condition": ["A", "B", "A", "B", "A", "B", "A", "B"],
            "Sex": ["Male", "Female", "Male", "Female", "Male", "Female", "Male", "Female"],
            "Time": [0, 0, 1, 1, 2, 2, 3, 3],
            "Value": [0.0, 10.0, 2.0, 12.0, 4.0, 14.0, 6.0, 16.0],
        }
    ).sort_values(["Condition", "Sex", "Time"]).reset_index(drop=True)

    actual = (
        _smooth_time_series(values, window=3)
        .sort_values(["Condition", "Sex", "Time"])
        .reset_index(drop=True)
    )

    pd.testing.assert_frame_equal(actual, expected)


def test_smooth_time_series_fails_when_timepoints_are_not_evenly_spaced():
    values = pd.DataFrame(
        {
            "Sample_ID": ["Sample_1", "Sample_2"],
            "Condition": ["A", "B"],
            "Sex": ["Male", "Female"],
            "0": [0.0, 10.0],
            "2": [6.0, 16.0],
            "3": [6.0, 16.0],
        }
    )

    with pytest.raises(AssertionError, match="Timepoints are not evenly spaced."):
        _smooth_time_series(values, window=3)
